lib/src/models/transaction_uploader.dart

```diff
@@ -26,198 +23,156 @@ const FATAL_CHUNK_UPLOAD_ERRORS = [
 ];
 
 class TransactionUploader {
-  int _chunkIndex = 0;
-  bool _txPosted = false;
-  int _lastRequestTimeEnd = 0;
-  int _totalErrors = 0;
-
-  int lastResponseStatus = 0;
-  String lastResponseError = '';
-
   final Transaction _transaction;
   final ArweaveApi _api;
-  final Random _random = Random();
+
+  bool get isComplete => _txPosted && uploadedChunks >= totalChunks;
+  int get totalChunks => _transaction.chunks!.chunks.length;
+  int get uploadedChunks => _uploadedChunks;
+
+  /// The progress of the current upload ranging from 0 to 1.
+  ///
+  /// Additionally accounts for the posting of the transaction header, therefore
+  /// data only uploads will start with a progress > 0.
+  double get progress =>
+      ((_txPosted ? 1 : 0) + uploadedChunks) / (1 + totalChunks);
+
+  final int maxConcurrentChunkUploadCount;
+
+  bool _txPosted = false;
+  int _uploadedChunks = 0;
 
   TransactionUploader(Transaction transaction, ArweaveApi api,
-      {bool forDataOnly = false})
+      {this.maxConcurrentChunkUploadCount = 128, bool forDataOnly = false})
```

Comment:
128 concurrency here causes an excessive amount of errors returned from the gateway during periods of network congestion. From our testing it seems that 32 is a safer default, especially given browser memory limitations

Regardless, this is a non-blocking comment now that it is injectable  (: